### ISSUE TITLE
[QST] how to do s4t s4t s32t GEMM efficiently?

### ISSUE THREAD

**Issue by Jacfger** [NVIDIA/cutlass#1711]
# Context:
I'm trying to write a PyTorch extension that does matrix multiplication of matrix A (MxK), B (KxN) where both matrix has type of 4 byte. (I.e. using `cutlass::int4b_t` as type). What frustrates me is that I realize there's no way to do GEMM A and B both represented in RowMajor fashion. But because how PyTorch works, both will be represented in RowMajor fashion, which leading to me having to makes matrix B pre-transposed before the multiplication done.

So I'm here to ask if it's possible to mitigate this issue, like for example, is there any way to do transpose efficiently (so to avoid matrix getting copied to-and-from global memory just because of doing the transpose)?
Because in my mind, I think there should be some way that during the phase where the matrix loading to the SHM, some kind of different loading could be done to load the data differently. So I don't quite understand why A, B can't be RowMajor, ColumnMajor at the same time.

# TL;DR
tl;dr, or to summarize, my questions are:
1. Why can't A, B be both in RowMajor or both in ColumnMajor when A and B are both int4b_t?
2. How can I do the transpose efficiently without affecting much of performance? Or is there any operations that I can add to GEMM so it automitcally transpose B before multiplication?

**Comment by thakkarV** [comment]
This is a limitation of the tensor core in ampere. You can transpose but it'll come at the cost of a lot of performance.

**Comment by Jacfger** [comment]
> This is a limitation of the tensor core in ampere. You can transpose but it'll come at the cost of a lot of performance.

I don't understand. Since cutlass load matrix in a hierarchical way, wouldn't it mitigate some of the performance drop? Or does the coalesence actually also matters a lot even when loading into SHM from global mem and to L1 cache from SHM?

Sorry if this is a stupid question..

**Comment by liuyt929** [comment]
excuse me. I've been writing a PyTorch extension recently, and I'd like to know that how you store your PyTorch matrix in 4-bit type, and how do you convert it directly into a form that cutlass gemm can use?  This question confused me a lot. thanks

**Comment by Jacfger** [comment]
> excuse me. I've been writing a PyTorch extension recently, and I'd like to know that how you store your PyTorch matrix in 4-bit type, and how do you convert it directly into a form that cutlass gemm can use?  This question confused me a lot. thanks
> 
> 

You can look into it the implementation of the paper QuaRot, or maybe just use their code.
